### ISSUE TITLE
`COREPACK_ENABLE_AUTO_PIN` should default to `0`

### ISSUE THREAD

**Issue by benmccann** [nodejs/corepack#485]
I don't want to use corepack. However, this setting means that every single contributor to my project will need to set `COREPACK_ENABLE_AUTO_PIN=0` and be careful not to accidentally check in the `packageManager` field since corepack will automatically add that field to force itself upon any project you use. This is even more annoying since pnpm can't be used without corepack once the `packageManager` field is present: https://github.com/pnpm/pnpm/issues/8087

**Comment by Lectem** [comment]
Same thing here.
I have projects that use pnpm, some don't, some require pinning the version and others don't.

I think nobody wants their package.json to be automatically modified on any command one could do. And even less on packages you do not own. 

I do NOT want to add this field on packages I do not own just because I contributed to it for a small fix or something.

Discoverability of such a feature is also bad, especially for juniors that will 100% fail to understand the implications of the new field and how to disable the feature.

**Comment by Lectem** [comment]
Note that it can (and should) in my opinion be fixed before the next version of node is released or the problem will most likely blow up.

For reference this was added by #413 in version 0.26. Original issue was #399. 

I think that a warning can be printed by default instead of directly modifying the `package.json`.

**Comment by AprilNEA** [comment]
I also believe that this should be disabled by default.

If we have to synchronize this environment variable within the team to prevent everyone from enabling it, that would be terrible.

**Comment by KaelWD** [comment]
At least allow it to be set in npmrc instead of an environment variable so public projects can force it off.

**Comment by vtnorton** [comment]
For me its another use case. I create a lot or demos, samples and help students with their code. I don't know the environment of each person, and I deal with a lot of different repositories and companies. 

Creating this `packageManager` by default, once I enabled corepack only on my machine, requires everyone to have the exact same package manager (including version) that I have on my computer. This leads to a awful developer experience for trying to running a demo, sample or a open source project for the first time. 

Yes, having an option to disable on `.npmrc` or something would be nice, but in most cases this leads to having more files then I honest need on my repo, that in my use cases I prefer to KISS. 

I did set `COREPACK_ENABLE_AUTO_PIN` to `0` so it doesn't botter me anymore, but I don't think this is the ideal solution.

**Comment by hardfist** [comment]
there're other monorepo tools like Rush which support pin `packageManager` version without using corepack, when run pnpm commands in these tools, the auto pin behavior may make inconsistent version than the version specified by monorepo tools.

**Comment by hyperknot** [comment]
I've run into the same issue. I've wrote about my case here, maybe someone also runs into this:
https://blog.hyperknot.com/p/corepacks-packagemanager-field

**Comment by ADTC** [comment]
Happy New Year! What is the response of maintainers on this?

Perhaps we could have an option of setting `"packageManager": null` which would effectively be equivalent of having `COREPACK_ENABLE_AUTO_PIN = 0` but **on a project level** instead of user level.

Committing that should effectively prevent corepack from ever writing the `packageManager` into `package.json` field and simply treat it the same as `COREPACK_ENABLE_AUTO_PIN = 0` in the project for every developer.

Unfortunately I suspect this won't be backwards compatible with older corepack versions which would just overwrite the current `packageManager` value (including `null`).

**Comment by MikeMcC399** [comment]
The default value of `COREPACK_ENABLE_AUTO_PIN` allowing Corepack to autopin is a barrier to adoption of Corepack in a mixed environment contributing to Yarn repos whilst still dealing with Yarn Classic repos which have not transitioned to Corepack.

If I execute `yarn -v` in a Yarn 1 Classic repo that I don't own and which has no `packageManager` field set in its `package.json` file, I don't want corepack to add one for me. As far as the repo is concerned, I would like `yarn -v` to be a read-only command. By default however, if corepack is enabled, it is not read-only.

I would like to have corepack permanently enabled for Yarn, however this situation means that I am enabling and disabling corepack instead as a workaround, which causes other difficulties, as the `corepack enable` / `corepack disable` sequence does not save and restore the links to any globally installed Yarn Classic version (see https://github.com/nodejs/corepack/issues/112). Yarn Classic is still there in the `node` directory, however it is no longer linked and essentially needs to be re-installed to get the link back.

I would also like to see  `COREPACK_ENABLE_AUTO_PIN` set to `0` by default, for the above reasons.

**Comment by ADTC** [comment]
Sidenote, just wondering why `packageManager` was not just a hint with `^` caret to indicate the minimum minor version instead of a pinned exact version? 

And why the SHA of exact version is not in the lock file, since that's what the lock file is for? 

Just like Node packages...

**Comment by MikeMcC399** [comment]
@hyperknot 

> Just to add - I can see that corepack is experimental and needs to be opted in - if this is the case why is it doing anything? I have upgraded to node 20 and I get this behaviour by default without enabling anything.

Would you like to open a new issue and say which operating system you are using and how you installed Node.js? As you say, Corepack is not supposed to be enabled by default.

**Comment by apieceofbart** [comment]
Yeah, sorry, I should have realized this is a completely different issue! I have already disabled corepack but will have folks with similar setup to mine so we'll try to get more info on the topic and move to new issue. I'm on macOS and using nvm to install node.

**Comment by jportner** [comment]
> Just to add - I can see that corepack is experimental and needs to be opted in - if this is the case why is it doing anything? I have upgraded to node 20 and I get this behaviour by default without enabling anything.

> Yeah, sorry, I should have realized this is a completely different issue! I have already disabled corepack but will have folks with similar setup to mine so we'll try to get more info on the topic and move to new issue. I'm on macOS and using nvm to install node.

@apieceofbart it looks like you didn't get around to opening a separate issue. I ran into this too, so I wanted to drop a line here in case other folks run into it too.

I never explicitly enabled Corepack, but when I upgraded my Node.js version, Corepack was unexpectedly enabled and it started pinning the package manager as described in this issue.

It turns out that `nvm` is automatically enabling Corepack when you use the `--reinstall-packages-from` argument. Perhaps you did the same? I filed a bug in the `nvm` repo: https://github.com/nvm-sh/nvm/issues/3544

**Comment by apieceofbart** [comment]
@jportner  thank you! I didn't follow up because it seems that no one else had that problem and I had trouble reproducing. 
You're spot on - I did run `--reinstall-packages-from` after bumping node via `nvm` so that's definitely the same case. Great job finding this out!

**Comment by trivikr** [comment]
This change should be made before 1.0.0 as per https://github.com/nodejs/corepack/issues/687#issuecomment-2746371851

**Comment by MikeMcC399** [comment]
@EseCurtis 

> where can i set my COREPACK_ENABLE_AUTO_PIN ? or how can i set it, thats the nquestion

See the documentation https://github.com/nodejs/corepack/blob/main/README.md#environment-variables

It's an environment variable, so how you set it depends on your operating system and your shell / terminal:
Linux / macOS: `export`
Windows cmd: `set`
Windows PowerShell `$env`
